### PR TITLE
ramips: add support for Mercusys MR70X

### PR DESCRIPTION
**Pull request by koreapyj**
Add support for Mercusys MR70X.

Hardware:
- SoC: MediaTek MT7621AT (880MHz, Duel-Core)
- RAM: DDR3 128MB
- Flash: Winbond W25Q128JV (SPI-NOR 16MB)
- WiFi: MediaTek MT7915D (2.4GHz, 5GHz, DBDC)
- Ethernet: MediaTek MT7530 (WAN x1, LAN x3, SoC)
- UART: [GND, RX, TX, 3.3V] (115200 8N1, J1)

Installation:
1. Flash factory image. This can be done using stock web ui
2. Connect to OpenWrt with an SSH connection to 192.168.1.1
3. Perform sysupgrade with sysupgrade image

Revert to stock firmware:
- Flash stock firmware via OEM Web UI Recovery mode

Web UI Recovery method:
1. Unplug the router
2. Plug in and hold reset button 5~10 secs
3. Set your computer IP address manually to 192.168.1.x / 255.255.255.0
4. Flash image with web browser to 192.168.1.1

### CODE REVIEW

**Comment by koreapyj** [comment]
Has MT7915 EEPROM issue.
```
[   50.086325] mt7915e_hif 0000:01:00.0: enabling device (0000 -> 0002)
[   50.104323] mt7915e 0000:02:00.0: enabling device (0000 -> 0002)
[   50.247121] mt7915e 0000:02:00.0: HW/SW Version: 0x8a108a10, Build Time: 20201105222230a
[   50.562240] mt7915e 0000:02:00.0: WM Firmware Version: ____000000, Build Time: 20201105222304
[   50.598479] mt7915e 0000:02:00.0: WA Firmware Version: DEV_000000, Build Time: 20201105222323
[   50.754974] mt7915e 0000:02:00.0: eeprom load fail, use default bin
[   50.761394] mt7915e 0000:02:00.0: Direct firmware load for mediatek/mt7915_eeprom_dbdc.bin failed with error -2
[   50.771467] mt7915e 0000:02:00.0: Falling back to sysfs fallback for: mediatek/mt7915_eeprom_dbdc.bin
[   50.881580] mt7915e: probe of 0000:02:00.0 failed with error -12
```

**Comment by koreapyj** [comment]
vendor GPL source code available at https://www.mercusys.com/en/download/mr70x#GPLCode

**Comment by nicefile** [comment]
if My table from MR1800X is same as MR70X
dev:    size   erasesize  name offset
mtd0: 00040000 00010000 "uboot" 00000000
mtd1: 00400000 00010000 "uImage" 00040000
mtd2: 00bb0000 00010000 "rootfs" 00440000
mtd3: 00010000 00010000 "ART" 00ff0000

FF0000 art (factory) is starting there
also found some OEM data under 00fa0000-00ff0000 (like MAC address)
Please check this https://www.openrouter.info/forum/viewtopic.php?f=19&t=4952 for DTS config (except SPI layout) 
I've used breed loader and extracted ART from oem SPI

**Comment by mrkiko** [comment]
Very nice to have offset values. Do you calculate them by hand or via a script or something?Thanks!

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+				};
+			};
+		};
+	};
+};
+
+&ethernet {
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii1_pins &mdio_pins>;
+};
```
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:94]: You'll not need this after 7774b8601937.

**Comment by nicefile** [comment]
@koreapyj Is this pull abandoned ? Have You checked my link and 4pda for safeloader  settings ?

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		};
+	};
+
+	// keys {
+	// 	compatible = "gpio-keys";
+
+	// 	reset {
+	// 		label = "reset";
+	// 		gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+	// 		linux,code = <KEY_RESTART>;
```
Comment (hauke) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:44]: why is this commented? Is this not working?

Comment (koreapyj) [reply target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:44]: It's just copied part but I haven't tested yet

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts (hunk 1 above)
Comment (koreapyj) [reply target/linux/ramips/dts/mt7621_mercusys_mr70x.dts]: removed.

**Comment by koreapyj** [comment]
Ready for review. [firmware-utils patch](https://patchwork.ozlabs.org/project/openwrt/list/?series=303280) should merged first.

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts (hunk 2 above)
Comment (koreapyj) [reply target/linux/ramips/dts/mt7621_mercusys_mr70x.dts]: resolved.

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_green: dome_green {
```
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:27]: The convention is dash for node names, underscore for label names.

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+			label = "green:dome";
```
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:28]: Sorry, could you explain what is "dome"?

File: target/linux/ramips/dts/mt7621_mercusys_mr70x.dts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+				reg = <0xfa0000 0x60000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "i2c";
```
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_mercusys_mr70x.dts:86]: "uart3" should be added due to the reset button (gpio8).

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network [hunk 6]
```diff
@@ -185,6 +186,11 @@ ramips_setup_macs()
 		wan_mac=$lan_mac
 		label_mac=$lan_mac
 		;;
+	mercusys,mr70x)
+		lan_mac=$(mtd_get_mac_binary factory 0x8)
+		wan_mac=$(macaddr_add "$lan_mac" 1)
+		label_mac=$lan_mac
+		;;
```
Comment (mans0n) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network:193]: In the first revision you had the MAC setup in the dts, why did you move it here?

**Review by mans0n — COMMENTED** [review]
Looks good, just a few nitpicks:

**Comment by mans0n** [comment]
> Has MT7915 EEPROM issue.

Just curious, what was the problem and how did you fix it?

**Comment by mans0n** [comment]
Just FYI, "[Mercusys](https://www.mercusys.com/)" is a new brand name for global market, and the domestic vendor name is still "[Mercury](https://www.mercurycom.com.cn/)". We have two Mercury devices in this repo, and [MW4530R](https://openwrt.org/toh/mercury/mw4530r_v1) is known to be a clone of TP-Link TL-WDR4310. No wonder why this device uses tplink-safeloader.

**Comment by mans0n** [comment]
> ramips: add support for Mercusys MR70X
> Add support for Mercusys MR70X.

Please insert a blank line right after the commit title.

**Comment by kajoj2** [comment]
Hello, 
I just bought Mercysys MR70X or MX1800 (this two names are on the box). 
I had tried to compile this PR but with defaults configs (no additional packages or kernel tweaks) but I'm getting error: 
```
WARNING: Image file /home/karol/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/tmp/openwrt-ramips-mt7621-mercusys_mr70x-squashfs-sysupgrade.bin is too big:  > 16121856

```

I've complied firmware for another router with this same amount of memory without any issue.

Can someone give me a hint what to do, this is a bug?

**Comment by DragonBluep** [comment]
@koreapyj Please take a look at this PR #10451

**Comment by amzamzamzamz** [comment]
> EDIT: it was a bug in master branch. I just complied and flashed everything works great. Thanks for your work @koreapyj !
>
HI. Can you explain what bug and how to fix it? The same problem

**Comment by kajoj2** [comment]
To be honest, I don't know where is bug.  But when I use master branch and apply this PR on top of it, fix it and I compiled with success.

**Comment by amzamzamzamz** [comment]
> To be honest, I don't know where is bug. But when I use master branch and apply this PR on top of it, fix it and I compiled with success.

So... Can you share sysupgrade bin?

**Comment by kajoj2** [comment]
[openwrt-ramips-mt7621-mercusys_mr70.zip](https://github.com/openwrt/openwrt/files/9483307/openwrt-ramips-mt7621-mercusys_mr70.zip)

Sure, I think you need factory, not sysupgrade. I uploaded all files. But with a couple extra packages, I hope you don't mind.

**Comment by amzamzamzamz** [comment]
> [openwrt-ramips-mt7621-mercusys_mr70.zip](https://github.com/openwrt/openwrt/files/9483307/openwrt-ramips-mt7621-mercusys_mr70.zip)
> 
> Sure, I think you need factory, not sysupgrade. I uploaded all files. But with a couple extra packages, I hope you don't mind.

Thanks
Wireless doesn't work. How to enable it?

**Comment by kajoj2** [comment]
For me, it works without any issue, maybe this is config issue? 

Unfortunately, I don't have telegram

**Comment by amzamzamzamz** [comment]
> For me, it works without any issue, maybe this is config issue?
> 
> Unfortunately, I don't have telegram

Hi

Can you share your "openwrt" folder content? 
I had tried to compile binary, but I'm getting error: 

I need to add some kernel modules to firmware.

**Comment by amzamzamzamz** [comment]
> For me, it works without any issue, maybe this is config issue?
> 
> Unfortunately, I don't have telegram

Buddy, i need your help.
I had tried to compile binary, but I'm getting error.
"WARNING: Image file /home/serg/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/tmp/openwrt-ramips-mt7621-mercusys_mr70x-squashfs-sysupgrade.bin is too big:  > 16121856"
How did you get so small binary size?
Can you share your "openwrt" folder content?
I need to add some kernel modules to firmware.

**Comment by creew** [comment]
> > For me, it works without any issue, maybe this is config issue?
> > Unfortunately, I don't have telegram
> 
> Buddy, i need your help. I had tried to compile binary, but I'm getting error. "WARNING: Image file /home/serg/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/tmp/openwrt-ramips-mt7621-mercusys_mr70x-squashfs-sysupgrade.bin is too big: > 16121856" How did you get so small binary size? Can you share your "openwrt" folder content? I need to add some kernel modules to firmware.

If you build from tagged version (not master), you must change version of firmware-utils in tools/firmware-utils/Makefile to actual from master

**Comment by KryptoSynth** [comment]
Hey, I couldn't access openwrt so i booted into the recovery ui and flashed the stock firmware, it says flashed successfully but the router does not reboot and seems to be stuck in recovery mode i can still access the recovery ui, any idea what i'm supposed to do?

I even tried flashing openwrt factory bin but got the same result.

**Comment by metropt** [comment]
Can anyone elaborate this step `Flash factory image. This can be done using stock web ui
` I may buy one to try openwrt. Thank you

**Comment by amzamzamzamz** [comment]
> Can anyone elaborate this step `Flash factory image. This can be done using stock web ui ` I may buy one to try openwrt. Thank you

HI. I have MR70X under openwrt. Everything works fine.
BUT!!! Now in stores they sell mainly the second version (v2) with Qualcomm hardware. Openwrt works only on v1. Be careful at shopping.
Hardware version you can see at bottom of retail box.

**Comment by amzamzamzamz** [comment]
> thank you very much

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network (hunk 6 above)
Comment (hauke) [reply target/linux/ramips/mt7621/base-files/etc/board.d/02_network:193]: I would also prefer if the mac addresses are configured in the dts file and not in this script where possible.

**Comment by forceintruth** [comment]
> Web UI Recovery method:
> 
>     1. Unplug the router
> 
>     2. Plug in and hold reset button 5~10 secs
> 
>     3. Set your computer IP address manually to 192.168.1.x / 255.255.255.0
> 
>     4. Flash image with web browser to 192.168.1.1

Hi! I follow the instructions, pressed the button, turned on the power, manually registered the IP, went to 192.168.1.1, uploaded my native firmware, tried all versions, and 1.0.0 too, the firmware is uploaded, and nothing happens, no reboot, nothing. If you manually pull out the power - the router is loaded into openvrt. Tell me please, can something be done after uploading the native firmware?

**Comment by Djfe** [comment]
> HI. I have MR70X under openwrt. Everything works fine. BUT!!! Now in stores they sell mainly the second version (v2) with Qualcomm hardware. Openwrt works only on v1. Be careful at shopping. Hardware version you can see at bottom of retail box.

Are you certain? Only the MR80X has a revision 2 on their website. And that device had a qualcomm in rev 1 already. So maybe you mistook an mr80x for an mr70x

**Comment by amzamzamzamz** [comment]
> 

You wrong. MR70X based on mediatek ramips cpu.
And yes, MR70X have two revisions - v1 (this thread), and quallcom based v2

**Comment by amzamzamzamz** [comment]
> > HI. I have MR70X under openwrt. Everything works fine. BUT!!! Now in stores they sell mainly the second version (v2) with Qualcomm hardware. Openwrt works only on v1. Be careful at shopping. Hardware version you can see at bottom of retail box.
> 
> Are you certain? Only the MR80X has a revision 2 on their website. And that device had a qualcomm in rev 1 already. So maybe you mistook an mr80x for an mr70x

You wrong. MR70X based on mediatek ramips cpu.
And yes, MR70X have two revisions - v1 (this thread), and quallcom based v2

**Comment by Djfe** [comment]
thank you for the info.
Do you know the actual SoC's name? Are there any specs out there?
mercusys doesn't tell us anything about a revision 2
https://www.mercusys.com/dk/product/details/mr70x/#specifications

File: target/linux/ramips/image/mt7621.mk [hunk 7]
```diff
@@ -1120,6 +1120,18 @@ define Device/mediatek_mt7621-eval-board
...
 TARGET_DEVICES += mediatek_mt7621-eval-board
 
+define Device/mercusys_mr70x
+  $(Device/dsa-migration)
+  $(Device/tplink-safeloader)
+  KERNEL := $(KERNEL_DTB) | uImage lzma
+  IMAGE_SIZE := 15744k
+  DEVICE_VENDOR := Mercusys
+  DEVICE_MODEL := MR70X
+  DEVICE_PACKAGES := kmod-mt7915e
```
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:1130]: Please replace `kmod-mt7915e` with `kmod-mt7915-firmware`. Reference: https://github.com/openwrt/openwrt/commit/fc9dd3f0834317ceb20700fd7033987695661097

**Comment by khagaroth** [comment]
> Do you know the actual SoC's name? Are there any specs out there? mercusys doesn't tell us anything about a revision 2

The Mercusys website is a mess, V2 is only listed on some EU sites. There is no firmware anywhere, but I did find GPL dump for V2 on the Mercusys UK site. I don't know where to look for the info on SOC, but after skimming through it quickly, there are some references to ipq5018 and  TP-Link AX55 in there.

**Comment by Djfe** [comment]
thanks for telling me, that local pages don't show all versions.
Here are all GPL Uploads:
https://www.mercusys.com/en/support/gpl-code?model=MR70X

**Comment by khagaroth** [comment]
> > Web UI Recovery method:
> > ```
> > 1. Unplug the router
> > 
> > 2. Plug in and hold reset button 5~10 secs
> > 
> > 3. Set your computer IP address manually to 192.168.1.x / 255.255.255.0
> > 
> > 4. Flash image with web browser to 192.168.1.1
> > ```
> 
> Hi! I follow the instructions, pressed the button, turned on the power, manually registered the IP, went to 192.168.1.1, uploaded my native firmware, tried all versions, and 1.0.1 too, the firmware is uploaded, and nothing happens, no reboot, nothing. If you manually pull out the power - the router is loaded into openwrt. Tell me please, can something be done after uploading the native firmware?

Not sure if you managed to fix it already, but for me the Web UI recovery didn't work because of the long file name of the firmware file (the original file name of the stock firmware is 64 characters long). After I renamed it to a short file name, it started working.

**Comment by forceintruth** [comment]
> Not sure if you managed to fix it already, but for me the Web UI recovery didn't work because of the long file name of the firmware file (the original file name of the stock firmware is 64 characters long). After I renamed it to a short file name, it started working.

Thank you for your reply. I returned to my native firmware using the CH341A programmer, but I will take into account your hint.

**Comment by nicefile** [comment]
@koreapyj Is this one is abandoned ? Next stable branch out is soon™  Can You please proceed with those suggested changes or close this so i can PR it from my branch.

**Comment by khagaroth** [comment]
If anyone is going to finish this, it would be nice to also look at the possibility to implement this change: https://github.com/openwrt/openwrt/commit/f1c9afd801380a05a91d979b475c76cc0a67caae

**Comment by nicefile** [comment]
> If anyone is going to finish this, it would be nice to also look at the possibility to implement this change: [f1c9afd](https://github.com/openwrt/openwrt/commit/f1c9afd801380a05a91d979b475c76cc0a67caae)

Yes I've created support [branch ](https://github.com/nicefile/openwrt-vector/tree/mr70x)with all those little suggestions . Just don't want to spam duplicated PR.

**Comment by kajoj2** [comment]
Thanks, @nicefile ! So it's time to upgrade my AP. 

I think current PR is dead and creator of it lose interest of it.

**Comment by nicefile** [comment]
@mans0n this PR is 9 months dead . How we can proceed forward ?

**Comment by khagaroth** [comment]
Err, he already did, read his message once again, there is a link to his uptodate PR.

**Comment by blocktrron** [comment]
Merged to my staging tree, thanks!

**Comment by tst0** [comment]
Hey, anybody can do this for MR80X?

**Comment by Djfe** [comment]
likely not possible. Please buy a supported mr70x or mr90x if you want openwrt.
